File: language/src/main/java/com/oracle/truffle/lama/nodes/expression/LamaDivNode.java
package com.oracle.truffle.lama.nodes.expression;

import com.oracle.truffle.api.dsl.Fallback;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.nodes.NodeInfo;
import com.oracle.truffle.lama.LamaException;
import com.oracle.truffle.lama.nodes.LamaBinaryNode;


@NodeInfo(shortName = "/")
public abstract class LamaDivNode extends LamaBinaryNode {

    @Specialization(rewriteOn = ArithmeticException.class)
    protected long div(long left, long right) {
        return Math.floorDiv(left, right);
    }

    @Fallback
    protected Object typeError(Object left, Object right) {
        throw LamaException.typeError(this, left, right);
    }

}

File: language/src/main/java/com/oracle/truffle/lama/nodes/expression/LamaDivNode.kt
package com.oracle.truffle.lama.nodes.expression

import com.oracle.truffle.api.dsl.Fallback
import com.oracle.truffle.api.dsl.Specialization
import com.oracle.truffle.api.nodes.NodeInfo
import com.oracle.truffle.lama.SLException
import com.oracle.truffle.lama.nodes.LamaBinaryNode

@NodeInfo(shortName = "/")
abstract class LamaDivNode : LamaBinaryNode() {
    @Specialization(rewriteOn = [ArithmeticException::class])
    @Throws(ArithmeticException::class)
    protected fun div(left: Long, right: Long): Long {
        val result = left / right
        if (left and right and result < 0) {
            throw ArithmeticException("long overflow")
        }
        return result
    }

    @Fallback
    protected fun typeError(left: Any?, right: Any?): Any {
        throw SLException.typeError(this, left, right)
    }
}